graphql_compiler/cost_estimation/cardinality_estimator.py

```diff
@@ -82,57 +82,126 @@ def _get_parent_and_child_name_from_edge(schema_graph, child_location):
     return parent_name_from_edge, child_name_from_edge
 
 
-def _estimate_children_per_parent(
-    schema_graph, statistics, query_metadata, parameters, child_location, parent_location
+def _get_outbound_and_inbound_names_from_locations(query_metadata, child_location, parent_location):
+    """Get the out and in classnames of a location and its parent from the last edge information."""
+    edge_direction, _ = _get_last_edge_direction_and_name_to_location(child_location)
+    parent_name_from_location = query_metadata.get_location_info(parent_location).type.name
+    child_name_from_location = query_metadata.get_location_info(child_location).type.name
+    if edge_direction == INBOUND_EDGE_DIRECTION:
+        outbound_vertex_name = child_name_from_location
+        inbound_vertex_name = parent_name_from_location
+    elif edge_direction == OUTBOUND_EDGE_DIRECTION:
+        outbound_vertex_name = parent_name_from_location
+        inbound_vertex_name = child_name_from_location
+    return outbound_vertex_name, inbound_vertex_name
+
+
+def _probe_statistics_for_edge_count_between_vertex_pair(
+    statistics, query_metadata, child_location, parent_location
 ):
-    """Estimate the number of edges per parent_location that connect to child_location vertices.
+    """Probe statistics for the count of edges connecting two vertex classes."""
+    _, edge_name = _get_last_edge_direction_and_name_to_location(child_location)
+    outbound_vertex_name, inbound_vertex_name = _get_outbound_and_inbound_names_from_locations(
+        query_metadata, child_location, parent_location
+    )
+    probe_result = statistics.get_edge_count_between_vertex_pair(
+        outbound_vertex_name, inbound_vertex_name, edge_name
+    )
+    return probe_result
 
-    Given a parent location of type A and child location of type B, assume all AB edges are
-    distributed evenly over A vertices, so the expected number of child edges per parent vertex is
-    (# of AB edges) / (# of A vertices). If A and B are subclasses of C and D respectively, we only
-    have access to CD edges, so in general, we'll use (# of CD edges) / (# of C vertices), but since
-    we're only interested in edges to Bs, we'll scale this result by the fraction of Bs over Ds.
+
+def _estimate_edge_count_between_vertex_pair_using_class_count(
+    schema_graph, statistics, query_metadata, child_location, parent_location
+):
+    """Estimate the count of edges connecting two vertex classes via the class_count statistic.
+
+    Given a parent location of type A and a child location of type B, we estimate the number of AB
+    edges using class counts. If A and B are subclasses of C and D respectively, we only have
+    access to CD edges, so in general, we assume that the number of CD edges with A as an endpoint
+    are proportional to the fraction of A vertices over C vertices, and likewise for B and D.
```

Comment:
Thoughts on rewording this part to make our assumption more precise? One possible rewording below, but feel free to suggest changes. 

From 

"we assume that the number of CD edges with A as an endpoint are proportional to the fraction of A vertices over C vertices, and likewise for B and D."

to

"we assume that the edges between a vertex $c$ of type C and vertex $d$ of type D are independent of whether $c$ is of type $A$ and whether $d$ is of type $B$. In other words, we assume the fraction of edges from $c$ to $d$ such that $c$ is of type A is the same as the fraction of C vertices in general that are also of type A. Likewise, we assume the fraction of edges from $c$ to $d$ such that $d$ is of type B is the same as the fraction of D vertices in general that are also of type B."

Reply:
> Thoughts on rewording this part to make our assumption more precise? One possible rewording below, but feel free to suggest changes.
> 
> From
> 
> "we assume that the number of CD edges with A as an endpoint are proportional to the fraction of A vertices over C vertices, and likewise for B and D."
> 
> to
> 
> "we assume that the edges between a vertex $c$ of type C and vertex $d$ of type D are independent of whether $c$ is of type $A$ and whether $d$ is of type $B$. In other words, we assume the fraction of edges from $c$ to $d$ such that $c$ is of type A is the same as the fraction of C vertices in general that are also of type A. Likewise, we assume the fraction of edges from $c$ to $d$ such that $d$ is of type B is the same as the fraction of D vertices in general that are also of type B."

I like the suggestion of more explicitly stating the assumption. Since it's very easy to explain this hand-wavingly, I'll try expressing it in form of an example.

Reply:
I've reworded it. I'd love to hear your thoughts on the current iteration.

Reply:
nit: "with A as an endpoint" -> "with an endpoint of class A"

everything else looks good to me!

Reply:
That's not the version I decided on, probably some git thing that I'll fix soon.